### PR TITLE
Update should_rank_save_checkpoint for multi nodes

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ananthsub** [comment]
@kaushikb11 - wondering if you took a look at https://github.com/PyTorchLightning/pytorch-lightning/issues/9074

IMO, this property should become an implementation detail of the training type plugin after the introduction of the checkpoint IO plugin

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -369,6 +369,20 @@ def configure_ddp(self) -> None:
...
+    @property
+    def should_rank_save_checkpoint(self) -> bool:
+        """Returns whether the checkpoint should be saved (rank based)"""
+        if self.num_nodes > 1:
+            sync_dir = None
+            if self.is_global_zero:
+                sync_dir = tempfile.mkdtemp()
+            self.barrier()
+            sync_dir = self.broadcast(sync_dir)
+            shared_file_system = self.all_gather(torch.tensor(int(os.path.exists(sync_dir)))).sum() == self.world_size
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:381]: the os.path.exists check will not work or a remote filesystem (s3, gcs) used with fsspec

**Comment by kaushikb11** [comment]
@ananthsub Thanks for creating the RFC, will take a look!

We wish to support the use cases for writing checkpoints for multi nodes when on separate as well as shared file systems.

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -369,6 +369,20 @@ def configure_ddp(self) -> None:
         )
         self._register_ddp_hooks()
 
+    @property
+    def should_rank_save_checkpoint(self) -> bool:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp.py:373]: Let's add a lru_cache.

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:373]: ddp_spawn needs this too.

**Comment by tchaton** [comment]
> @kaushikb11 - is there an issue describing the current problem with DDP & checkpoint saving?
> 
> also i'm wondering if you took a look at #9074

Hey Ananth, in the case where your are running in a multi node system with a non-shared file system, then we need to do the follow:

* For Sharded Plugin: Save / load on all ranks
* For Non-Sharded Plugin like DDP, save / load on all local_zero.

**Comment by tchaton** [comment]
Hey @kaushikb11.

Any updates ?

Best,
T.C
